File: backend/src/routes/auth.routes.js
/**
 * Auth Routes — Login, logout, profile
 * POST /api/v1/auth/login
 * GET  /api/v1/auth/me
 * POST /api/v1/auth/logout
 */
const router = require('express').Router();
const { supabase } = require('../config/supabase');
const { requireAuth } = require('../middleware/auth.middleware');
const { requirePermission, ROLES } = require('../middleware/role.guard');

// ── Login ─────────────────────────────────────────────────────────────────
router.post('/login', async (req, res) => {
  try {
    const { email, password } = req.body;
    if (!email || !password) {
      return res.status(400).json({ error: 'Email and password required' });
    }

    // Authenticate with Supabase Auth
    const { data, error } = await supabase.auth.signInWithPassword({ email, password });
    if (error) return res.status(401).json({ error: 'Invalid email or password' });

    const { user, session } = data;

    // Fetch role profile
    const { data: profile } = await supabase
      .from('profiles')
      .select('id, role, branch_id, full_name, status')
      .eq('auth_uid', user.id)
      .single();

    if (!profile || profile.status === 0) {
      return res.status(403).json({ error: 'Account is inactive or not found' });
    }

    // Return token + role info — Flutter app stores this
    res.json({
      success: true,
      access_token: session.access_token,
      refresh_token: session.refresh_token,
      user: {
        id:        profile.id,
        role:      profile.role,
        name:      profile.full_name,
        branch_id: profile.branch_id,
        email:     user.email,
      }
    });
  } catch (err) {
    console.error('Login error:', err.message);
    res.status(500).json({ error: 'Login failed' });
  }
});

// ── Send OTP (Email based) ────────────────────────────────────────────────
router.post('/send-otp', async (req, res) => {
  try {
    const { email } = req.body;
    if (!email) return res.status(400).json({ error: 'Email required' });

    // Send OTP via Supabase
    const { error } = await supabase.auth.signInWithOtp({
      email,
      options: { shouldCreateUser: false }, // Prevent random signups
    });
    
    if (error) return res.status(400).json({ error: error.message });

    res.json({ success: true, message: 'OTP sent successfully to email' });
  } catch (err) {
    console.error('OTP Send error:', err.message);
    res.status(500).json({ error: 'Failed to send OTP' });
  }
});

// ── Verify OTP ────────────────────────────────────────────────────────────
router.post('/verify-otp', async (req, res) => {
  try {
    const { email, otp } = req.body;
    if (!email || !otp) {
      return res.status(400).json({ error: 'Email and OTP required' });
    }

    // Verify OTP with Supabase Auth
    const { data, error } = await supabase.auth.verifyOtp({
      email,
      token: otp,
      type: 'email',
    });
    
    if (error) return res.status(401).json({ error: 'Invalid or expired OTP' });

    const { user, session } = data;

    // Fetch role profile
    const { data: profile } = await supabase
      .from('profiles')
      .select('id, role, branch_id, full_name, status')
      .eq('auth_uid', user.id)
      .single();

    if (!profile || profile.status === 0) {
      return res.status(403).json({ error: 'Account is inactive or not found' });
    }

    // Return token + role info
    res.json({
      success: true,
      access_token: session.access_token,
      refresh_token: session.refresh_token,
      user: {
        id:        profile.id,
        role:      profile.role,
        name:      profile.full_name,
        branch_id: profile.branch_id,
        email:     user.email,
      }
    });
  } catch (err) {
    console.error('Verify OTP error:', err.message);
    res.status(500).json({ error: 'Failed to verify OTP' });
  }
});

// ── Get current user profile ───────────────────────────────────────────────
router.get('/me', requireAuth, async (req, res) => {
  try {
    const role = req.role;
    let extraData = {};

    if (role === 'student') {
      // Fetch student record linked to this profile
      const { data: student } = await supabase
        .from('students')
        .select('id, name, reg_no, roll_no, photo_url, course_id, branch_id, status')
        .eq('profile_id', req.profileId)
        .single();
      extraData = { student };
    }

    if (role === 'teacher') {
      const { data: employee } = await supabase
        .from('employees')
        .select('id, name, designation, department, branch_id')
        .eq('profile_id', req.profileId)
        .single();
      extraData = { employee };
    }

    res.json({
      success: true,
      profile: {
        id:       req.profileId,
        role:     req.role,
        name:     req.profile.full_name,
        branch_id: req.branchId,
        email:    req.user.email,
      },
      ...extraData
    });
  } catch (err) {
    res.status(500).json({ error: 'Could not fetch profile' });
  }
});

// ── Refresh token ─────────────────────────────────────────────────────────
router.post('/refresh', async (req, res) => {
  try {
    const { refresh_token } = req.body;
    if (!refresh_token) return res.status(400).json({ error: 'refresh_token required' });

    const { data, error } = await supabase.auth.refreshSession({ refresh_token });
    if (error) return res.status(401).json({ error: 'Token refresh failed' });

    res.json({
      success: true,
      access_token: data.session.access_token,
      refresh_token: data.session.refresh_token,
    });
  } catch (err) {
    res.status(500).json({ error: 'Refresh failed' });
  }
});

// ── Logout ────────────────────────────────────────────────────────────────
router.post('/logout', requireAuth, async (req, res) => {
  await supabase.auth.signOut();
  res.json({ success: true, message: 'Logged out' });
});

// ── Admin: Register Branch Admin ──────────────────────────────────────────
router.post('/admin/register-branch-admin', requireAuth, requirePermission('REGISTER_BRANCH_ADMIN'), async (req, res) => {
  const { email, password, name } = req.body;
  
  if (!email || !password || !name) {
    return res.status(400).json({ error: 'Email, password and name are required' });
  }

  try {
    // 1. Create user in Supabase Auth (Service Role bypasses signup confirmation if configured)
    const { data: authData, error: authError } = await supabase.auth.admin.createUser({
      email,
      password,
      email_confirm: true,
      user_metadata: { name }
    });

    if (authError) return res.status(400).json({ error: authError.message });

    const user = authData.user;

    // 2. Create Profile
    const { data: profile, error: profileError } = await supabase
      .from('profiles')
      .insert({
        auth_uid: user.id,
        full_name: name,
        role: ROLES.BRANCH_ADMIN,
        status: 1 // Active
      })
      .select()
      .single();

    if (profileError) {
      // Cleanup: delete auth user if profile creation fails
      await supabase.auth.admin.deleteUser(user.id);
      return res.status(500).json({ error: 'Failed to create user profile' });
    }

    // 3. Create Admin record
    const { error: adminError } = await supabase
      .from('admins')
      .insert({
        user_id: user.id,
        email: email,
        name: name
      });

    if (adminError) {
      console.warn('⚠️ Admin record creation failed, but user and profile exist:', adminError.message);
    }

    res.json({
      success: true,
      message: 'Branch Admin registered successfully',
      user: {
        id: profile.id,
        email: user.email,
        name: name
      }
    });
  } catch (err) {
    console.error('Admin Registration Error:', err.message);
    res.status(500).json({ error: 'Internal server error during registration' });
  }
});

// ── Admin: Register Teacher ───────────────────────────────────────────────
router.post('/admin/register-teacher', requireAuth, requirePermission('REGISTER_TEACHER'), async (req, res) => {
  const { email, password, name } = req.body;
  let branchId = req.body.branch_id;
  
  if (req.role === ROLES.BRANCH_ADMIN) {
    branchId = req.branchId; // Force their own branch
  }

  if (!email || !password || !name) {
    return res.status(400).json({ error: 'Email, password and name are required' });
  }

  try {
    // 1. Create user in Supabase Auth
    const { data: authData, error: authError } = await supabase.auth.admin.createUser({
      email,
      password,
      email_confirm: true,
      user_metadata: { name }
    });

    if (authError) return res.status(400).json({ error: authError.message });

    const user = authData.user;

    // 2. Create Profile
    const { data: profile, error: profileError } = await supabase
      .from('profiles')
      .insert({
        auth_uid: user.id,
        full_name: name,
        role: ROLES.TEACHER,
        branch_id: branchId,
        status: 1 // Active
      })
      .select()
      .single();

    if (profileError) {
      await supabase.auth.admin.deleteUser(user.id);
      return res.status(500).json({ error: 'Failed to create user profile' });
    }

    // 3. Create Employee record
    const { error: employeeError } = await supabase
      .from('employees')
      .insert({
        profile_id: profile.id,
        name: name,
        designation: 'Teacher',
        department: 'Academic',
        branch_id: branchId
      });

    if (employeeError) {
      console.warn('⚠️ Employee record creation failed, but user and profile exist:', employeeError.message);
    }

    res.json({
      success: true,
      message: 'Teacher registered successfully',
      user: {
        id: profile.id,
        email: user.email,
        name: name
      }
    });
  } catch (err) {
    console.error('Teacher Registration Error:', err.message);
    res.status(500).json({ error: 'Internal server error during registration' });
  }
});

// ── Super Admin ONLY: Reset Any User Password ──────────────────────────────
// requirePermission('RESET_USER_PASSWORD') blocks ALL non-super_admin roles before reaching this handler.
router.post('/admin/reset-password', requireAuth, requirePermission('RESET_USER_PASSWORD'), async (req, res) => {
  const { targetProfileId, newPassword } = req.body;

  if (!targetProfileId || !newPassword) {
    return res.status(400).json({ error: 'Target profile ID and new password are required' });
  }

  if (newPassword.length < 6) {
    return res.status(400).json({ error: 'Password must be at least 6 characters long' });
  }

  try {
    // 1. Fetch target profile
    const { data: targetProfile, error: profileError } = await supabase
      .from('profiles')
      .select('id, role, branch_id, auth_uid, full_name')
      .eq('id', targetProfileId)
      .single();

    if (profileError || !targetProfile) {
      return res.status(404).json({ error: 'Target user profile not found' });
    }

    // 2. Super Admin can reset ANY user's password — no further checks needed.
    // (requirePermission('RESET_USER_PASSWORD') already ensures only super_admin reaches here)
    // Prevent super_admin from accidentally being passed invalid targets via extra guard:
    if (!targetProfile.auth_uid) {
      return res.status(400).json({ error: 'Target user has no auth account linked. Cannot reset password.' });
    }

    // 3. Update password via Supabase Admin Auth API
    const { error: authError } = await supabase.auth.admin.updateUserById(
      targetProfile.auth_uid,
      { password: newPassword }
    );

    if (authError) {
      return res.status(400).json({ error: authError.message });
    }

    res.json({
      success: true,
      message: `Password updated successfully for ${targetProfile.full_name}`,
    });

  } catch (err) {
    console.error('Admin Password Reset Error:', err.message);
    res.status(500).json({ error: 'Internal server error during password reset' });
  }
});

module.exports = router;

File: dart_backend/lib/routes/auth_routes.dart
// lib/routes/auth_routes.dart
// Auth routes: login, register, OTP, logout, profile, admin registration, password reset
// Ported from auth.routes.js with BUG-2 and BUG-6 fixes applied.

import 'dart:convert';
import 'package:shelf/shelf.dart';
import 'package:shelf_router/shelf_router.dart';
import '../config/supabase_service.dart';
import '../middleware/auth_middleware.dart';
import '../middleware/role_guard.dart';
import '../middleware/rate_limiter.dart';
import '../models/user_session.dart';

Router buildAuthRouter() {
  final router = Router();

  // ── Public Student Register ───────────────────────────────────────────────
  router.post('/register', (Request req) async {
    try {
      final body = await _body(req);
      final email    = body['email'] as String?;
      final password = body['password'] as String?;
      final name     = body['name'] as String?;
      final phone    = body['phone'] as String?;
      final fatherName = body['father_name'] as String?;
      final dob      = body['dob'] as String?;
      final address  = body['address'] as String?;
      final gender   = body['gender'] as String?;
      final courseId = body['course_id'];
      final branchId = body['branch_id'];

      if (email == null || password == null || name == null) {
        return _json(400, {'error': 'Email, password and name are required'});
      }

      // Validate branch if provided
      if (branchId != null) {
        final branches = await SupabaseService.select(
          'branches',
          columns: 'id,status',
          filters: {'id': int.parse(branchId.toString())},
        );
        if (branches.isEmpty) return _json(400, {'error': 'Invalid branch ID'});
        if (branches.first['status'] != 1) {
          return _json(400, {'error': 'Selected branch is currently inactive'});
        }
      }

      // 1. Create auth user (auto-confirms email)
      final authUser = await SupabaseService.adminCreateUser(
        email: email,
        password: password,
        userMetadata: {'name': name, 'mobile': phone},
      );
      final authUid = authUser!['id'] as String;

      // 2. Create Profile record (role = student, status = 0 Pending)
      Map<String, dynamic> profile;
      try {
        profile = await SupabaseService.insert('profiles', {
          'auth_uid': authUid,
          'full_name': name,
          'role': roleStudent,
          'contact': phone,
          'email': email,
          'branch_id': branchId != null ? int.parse(branchId.toString()) : null,
          'status': 0, // Pending approval
          'permissions': [
            'READ_OWN_PROFILE', 'READ_OWN_FEES', 'READ_OWN_ATTENDANCE',
            'READ_OWN_MARKSHEET', 'READ_OWN_CERTIFICATE', 'READ_OWN_IDCARD', 'TAKE_EXAM'
          ],
        });
      } catch (e) {
        await SupabaseService.adminDeleteUser(authUid);
        return _json(500, {'error': 'Failed to create student profile: $e'});
      }

      // 3. Create Student record
      // BUG-2 FIX: use profile['id'] (profiles.id), NOT authUid (auth.users.id)
      try {
        await SupabaseService.insert('students', {
          'profile_id': profile['id'], // ← FIXED: was authUid in JS bug
          'name': name,
          'email': email,
          'contact': phone,
          'status': 0,
          'father_name': fatherName,
          'dob': dob,
          'address': address,
          'gender': gender,
          'course_id': courseId != null ? int.parse(courseId.toString()) : null,
          'branch_id': branchId != null ? int.parse(branchId.toString()) : null,
        });
      } catch (e) {
        print('⚠️ Student record creation warning: $e');
      }

      return _json(200, {
        'success': true,
        'message': 'Student registered successfully and is pending approval',
        'user': {'id': profile['id'], 'email': email, 'name': name},
      });
    } catch (e) {
      print('Student Registration Error: $e');
      return _json(500, {'error': 'Internal server error during registration'});
    }
  });

  // ── Login ─────────────────────────────────────────────────────────────────
  router.post('/login', (Request req) async {
    try {
      final body = await _body(req);
      final email    = body['email'] as String?;
      final password = body['password'] as String?;

      if (email == null || password == null) {
        return _json(400, {'error': 'Email and password required'});
      }

      final authData = await SupabaseService.signInWithPassword(
        email: email,
        password: password,
      );
      if (authData == null) {
        return _json(401, {'error': 'Invalid email or password'});
      }

      final accessToken  = authData['access_token'] as String;
      final refreshToken = authData['refresh_token'] as String;
      final user         = authData['user'] as Map<String, dynamic>;

      // Fetch role profile
      final profiles = await SupabaseService.select(
        'profiles',
        columns: 'id,role,branch_id,full_name,status,permissions',
        filters: {'auth_uid': user['id']},
      );

      if (profiles.isEmpty) {
        return _json(403, {'error': 'Account is inactive or not found'});
      }

      final profile = profiles.first;

      // BUG-6 FIX: use != 1 (not == 0) — catches null, suspended, etc.
      if (profile['status'] != 1) {
        return _json(403, {'error': 'Account is inactive or not found'});
      }

      return _json(200, {
        'success': true,
        'access_token': accessToken,
        'refresh_token': refreshToken,
        'user': {
          'id':          profile['id'],
          'role':        profile['role'],
          'name':        profile['full_name'],
          'branch_id':   profile['branch_id'],
          'email':       user['email'],
          'permissions': profile['permissions'] ?? [],
        },
      });
    } catch (e) {
      print('Login error: $e');
      return _json(500, {'error': 'Login failed'});
    }
  });

  // ── Send OTP ──────────────────────────────────────────────────────────────
  router.post('/send-otp', (Request req) async {
    try {
      final body  = await _body(req);
      final email = body['email'] as String?;
      if (email == null) return _json(400, {'error': 'Email required'});

      final ok = await SupabaseService.sendOtp(email);
      if (!ok) return _json(400, {'error': 'Failed to send OTP'});

      return _json(200, {'success': true, 'message': 'OTP sent successfully to email'});
    } catch (e) {
      return _json(500, {'error': 'Failed to send OTP'});
    }
  });

  // ── Verify OTP ────────────────────────────────────────────────────────────
  router.post('/verify-otp', (Request req) async {
    try {
      final body  = await _body(req);
      final email = body['email'] as String?;
      final otp   = body['otp'] as String?;
      if (email == null || otp == null) {
        return _json(400, {'error': 'Email and OTP required'});
      }

      final data = await SupabaseService.verifyOtp(email: email, token: otp);
      if (data == null) return _json(401, {'error': 'Invalid or expired OTP'});

      final user         = data['user'] as Map<String, dynamic>;
      final accessToken  = data['access_token'] as String;
      final refreshToken = data['refresh_token'] as String;

      final profiles = await SupabaseService.select(
        'profiles',
        columns: 'id,role,branch_id,full_name,status',
        filters: {'auth_uid': user['id']},
      );

      if (profiles.isEmpty || profiles.first['status'] != 1) {
        return _json(403, {'error': 'Account is inactive or not found'});
      }

      final profile = profiles.first;
      return _json(200, {
        'success': true,
        'access_token': accessToken,
        'refresh_token': refreshToken,
        'user': {
          'id': profile['id'], 'role': profile['role'],
          'name': profile['full_name'], 'branch_id': profile['branch_id'],
          'email': user['email'],
        },
      });
    } catch (e) {
      return _json(500, {'error': 'Failed to verify OTP'});
    }
  });

  // ── Get current user profile ──────────────────────────────────────────────
  router.get('/me', Pipeline()
      .addMiddleware(requireAuth())
      .addHandler((Request req) async {
    try {
      final session = req.context['session'] as UserSession;
      Map<String, dynamic> extraData = {};

      if (session.role == roleStudent) {
        final students = await SupabaseService.select(
          'students',
          columns: 'id,name,reg_no,roll_no,photo_url,course_id,branch_id,status',
          filters: {'profile_id': session.profileId},
        );
        extraData['student'] = students.isNotEmpty ? students.first : null;
      }

      if (session.role == roleTeacher) {
        final employees = await SupabaseService.select(
          'employees',
          columns: 'id,name,designation,department,branch_id',
          filters: {'profile_id': session.profileId},
        );
        extraData['employee'] = employees.isNotEmpty ? employees.first : null;
      }

      return _json(200, {
        'success': true,
        'profile': {
          'id': session.profileId, 'role': session.role,
          'name': session.fullName, 'branch_id': session.branchId,
          'email': session.profile['email'],
          'permissions': session.profile['permissions'] ?? [],
        },
        ...extraData,
      });
    } catch (e) {
      return _json(500, {'error': 'Could not fetch profile'});
    }
  }));

  // ── Refresh token ─────────────────────────────────────────────────────────
  router.post('/refresh', (Request req) async {
    try {
      final body         = await _body(req);
      final refreshToken = body['refresh_token'] as String?;
      if (refreshToken == null) return _json(400, {'error': 'refresh_token required'});

      final data = await SupabaseService.refreshSession(refreshToken);
      if (data == null) return _json(401, {'error': 'Token refresh failed'});

      return _json(200, {
        'success': true,
        'access_token':  data['access_token'],
        'refresh_token': data['refresh_token'],
      });
    } catch (e) {
      return _json(500, {'error': 'Refresh failed'});
    }
  });

  // ── Logout ────────────────────────────────────────────────────────────────
  router.post('/logout', Pipeline()
      .addMiddleware(requireAuth())
      .addHandler((Request req) async {
    try {
      final token = req.context['token'] as String;
      await SupabaseService.signOut(token);
      return _json(200, {'success': true, 'message': 'Logged out'});
    } catch (e) {
      return _json(500, {'error': 'Failed to log out from server'});
    }
  }));

  // ── Admin: Register Branch Admin ──────────────────────────────────────────
  router.post('/admin/register-branch-admin', Pipeline()
      .addMiddleware(requireAuth())
      .addMiddleware(requirePermission('REGISTER_BRANCH_ADMIN'))
      .addHandler((Request req) async {
    try {
      final body = await _body(req);
      final email    = body['email'] as String?;
      final password = body['password'] as String?;
      final name     = body['name'] as String?;

      if (email == null || password == null || name == null) {
        return _json(400, {'error': 'Email, password and name are required'});
      }

      final authUser = await SupabaseService.adminCreateUser(
        email: email, password: password,
        userMetadata: {'name': name},
      );
      final authUid = authUser!['id'] as String;

      Map<String, dynamic> profile;
      try {
        profile = await SupabaseService.insert('profiles', {
          'auth_uid': authUid,
          'full_name': name,
          'role': roleBranchAdmin,
          'status': 1,
          'email': email,
          'permissions': [
            'READ_OWN_PROFILE', 'MARK_ATTENDANCE', 'READ_BRANCH_STUDENTS',
            'UPLOAD_MARKS', 'ENROLL_STUDENT', 'RECORD_FEE', 'SUBMIT_MARKSHEET',
            'ISSUE_ADMIT_CARD', 'READ_BRANCH_FEES', 'MANAGE_NOTICES',
            'ISSUE_CERTIFICATE', 'SETUP_OWN_BRANCH', 'REGISTER_TEACHER',
          ],
        });
      } catch (e) {
        await SupabaseService.adminDeleteUser(authUid);
        return _json(500, {'error': 'Failed to create user profile'});
      }

      return _json(200, {
        'success': true,
        'message': 'Branch Admin registered successfully',
        'user': {'id': profile['id'], 'email': email, 'name': name},
      });
    } catch (e) {
      print('Admin Registration Error: $e');
      return _json(500, {'error': 'Internal server error during registration'});
    }
  }));

  // ── Admin: Register Teacher ───────────────────────────────────────────────
  router.post('/admin/register-teacher', Pipeline()
      .addMiddleware(requireAuth())
      .addMiddleware(requirePermission('REGISTER_TEACHER'))
      .addHandler((Request req) async {
    try {
      final session  = req.context['session'] as UserSession;
      final body     = await _body(req);
      final email    = body['email'] as String?;
      final password = body['password'] as String?;
      final name     = body['name'] as String?;
      // Force branch_admin to use their own branch
      final branchId = session.role == roleBranchAdmin
          ? session.branchId
          : body['branch_id'] as int?;

      if (email == null || password == null || name == null) {
        return _json(400, {'error': 'Email, password and name are required'});
      }

      final authUser = await SupabaseService.adminCreateUser(
        email: email, password: password,
        userMetadata: {'name': name},
      );
      final authUid = authUser!['id'] as String;

      Map<String, dynamic> profile;
      try {
        profile = await SupabaseService.insert('profiles', {
          'auth_uid': authUid,
          'full_name': name,
          'role': roleTeacher,
          'branch_id': branchId,
          'status': 1,
          'email': email,
          'permissions': [
            'READ_OWN_PROFILE', 'MARK_ATTENDANCE',
            'READ_BRANCH_STUDENTS', 'UPLOAD_MARKS',
          ],
        });
      } catch (e) {
        await SupabaseService.adminDeleteUser(authUid);
        return _json(500, {'error': 'Failed to create user profile'});
      }

      try {
        await SupabaseService.insert('employees', {
          'profile_id': profile['id'],
          'name': name,
          'designation': 'Teacher',
          'department': 'Academic',
          'branch_id': branchId,
        });
      } catch (e) {
        print('⚠️ Employee record creation failed: $e');
      }

      return _json(200, {
        'success': true,
        'message': 'Teacher registered successfully',
        'user': {'id': profile['id'], 'email': email, 'name': name},
      });
    } catch (e) {
      print('Teacher Registration Error: $e');
      return _json(500, {'error': 'Internal server error during registration'});
    }
  }));

  // ── Super Admin: Reset Any User Password ─────────────────────────────────
  router.post('/admin/reset-password', Pipeline()
      .addMiddleware(requireAuth())
      .addMiddleware(requirePermission('RESET_USER_PASSWORD'))
      .addHandler((Request req) async {
    try {
      final body            = await _body(req);
      final targetProfileId = body['targetProfileId'] as String?;
      final newPassword     = body['newPassword'] as String?;

      if (targetProfileId == null || newPassword == null) {
        return _json(400, {'error': 'Target profile ID and new password are required'});
      }
      if (newPassword.length < 8) {
        // SEC-4 FIX: raised from 6 to 8 chars
        return _json(400, {'error': 'Password must be at least 8 characters long'});
      }

      final profiles = await SupabaseService.select(
        'profiles',
        columns: 'id,role,branch_id,auth_uid,full_name',
        filters: {'id': targetProfileId},
      );
      if (profiles.isEmpty) {
        return _json(404, {'error': 'Target user profile not found'});
      }

      final target = profiles.first;
      final authUid = target['auth_uid'] as String?;
      if (authUid == null) {
        return _json(400, {'error': 'Target user has no auth account linked.'});
      }

      await SupabaseService.adminUpdateUserById(authUid, password: newPassword);

      return _json(200, {
        'success': true,
        'message': 'Password updated successfully for ${target['full_name']}',
      });
    } catch (e) {
      print('Admin Password Reset Error: $e');
      return _json(500, {'error': 'Internal server error during password reset'});
    }
  }));

  return router;
}

// ── Helpers ───────────────────────────────────────────────────────────────────
Future<Map<String, dynamic>> _body(Request req) async {
  final str = await req.readAsString();
  if (str.isEmpty) return {};
  return jsonDecode(str) as Map<String, dynamic>;
}

Response _json(int status, Map<String, dynamic> body) =>
    Response(status,
        body: jsonEncode(body),
        headers: {'Content-Type': 'application/json'});
